pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -889,6 +890,26 @@ protected void doTakeSnapshot() {
         numUnchangedSegments++;
         continue;
       }
+      // Try to acquire the segmentLock when taking snapshot for the segment because the segment directory can be
+      // modified, e.g. a new snapshot file can be added to the directory. If not taking the lock, the Helix task
+      // thread replacing the segment could fail. For example, we found FileUtils.cleanDirectory() failed due to
+      // DirectoryNotEmptyException because a new snapshot file got added into the segment directory just between two
+      // major cleanup steps in the cleanDirectory() method.
+      String segmentName = segment.getSegmentName();
+      Lock segmentLock = tableDataManager.getSegmentLock(segmentName);
+      boolean locked = segmentLock.tryLock();
+      if (!locked) {
+        // Try to get the segmentLock in a non-blocking manner to avoid deadlock. The Helix task thread takes
+        // segmentLock first and then the snapshot RLock when replacing a segment. However, the consuming thread has
+        // already acquired the snapshot WLock when reaching here, and if it has to wait for segmentLock, it may
+        // enter deadlock with the Helix task threads waiting for snapshot RLock.
+        // If we can't get the segmentLock, we'd better skip taking snapshot for the tracked segment. Because the
+        // validDocIds of the tracked segment might be wrong for the new segment. It's better to have no snapshot
+        // file on disk than having a wrong one.
+        // The numMissedSegments metrics below can help monitor such cases.
+        _logger.warn("Could not get segmentLock to take snapshot for segment: {}, skipping", segmentName);
+        continue;
```

Comment:
Instead of skipping should we return from here to ensure disjoint valid doc ids on disk? A similar conversation we had here: https://github.com/apache/pinot/pull/13285/files#r1635324871

Reply:
Good call. I thought on this but was mainly on replacing segment.

When replacing segment, the new segment folder won't have snapshot file as it's a newly uncompressed segment folder, so snapshot files put on disk after taking this round of snapshots should continue to be disjoint.

But this seems not right for reloading segment, as the snapshot file is copied back in to the segment folder when reloading.

I'll do this:
1. for segments with existing snapshot files, we don't need to abort early upon failure to to get lock, because new snapshot files only get smaller number of valid doc ids.
2. (new check) if we skipped any segment with existing snapshot files in the 1st for-loop, we need to skip the whole 2nd for-loop, as putting any new snapshot files on disk may lead to overlap
3. if we didn't skip any segment with existing snapshot files, we can do the 2nd for-loop and no need to abort early upon failure to get lock.

Reply:
Sounds good to me! Also we can change the logic in 2nd-for-loop to skip a segment for any error scenarios instead of throwing exception because now we will be ensuring that all the ones with snapshot files are snapshotted again and disjoint nature will be true. Wdyt?

Reply:
Refined, basically, we skip 2nd for loop if we skipped any segment in 1st for-loop either due to failure to get lock or failure to persist snapshot. And no need to abort too early upon failure, we can skip segments accordingly.